libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py

```diff
@@ -60,19 +93,369 @@ def validate_environment(  # pylint: disable=no-self-argument
                 "Could not import ads python package. "
                 "Please install it with `pip install oracle_ads`."
             ) from ex
+
         if not values.get("auth", None):
             values["auth"] = ads.common.auth.default_signer()
+
         values["endpoint"] = get_from_dict_or_env(
             values,
             "endpoint",
             "OCI_LLM_ENDPOINT",
         )
         return values
 
+    def _headers(self, is_async=False, body=None) -> Dict:
+        """Construct and return the headers for a request.
+
+        Args:
+            is_async (bool, optional): Indicates if the request is asynchronous.
+                Defaults to `False`.
+            body (optional): The request body to be included in the headers if
+                the request is asynchronous.
+
+        Returns:
+            Dict: A dictionary containing the appropriate headers for the request.
+        """
+        if is_async:
+            signer = self.auth["signer"]
+            req = requests.Request("POST", self.endpoint, json=body)
+            req = req.prepare()
+            req = signer(req)
+            headers = {}
+            for key, value in req.headers.items():
+                headers[key] = value
+
+            if self.streaming:
+                headers.update(
+                    {"enable-streaming": "true", "Accept": "text/event-stream"}
+                )
+            return headers
+
+        return (
+            {
+                "Content-Type": DEFAULT_CONTENT_TYPE_JSON,
+                "enable-streaming": "true",
+                "Accept": "text/event-stream",
+            }
+            if self.streaming
+            else {
+                "Content-Type": DEFAULT_CONTENT_TYPE_JSON,
+            }
+        )
+
+    def completion_with_retry(
+        self, run_manager: Optional[CallbackManagerForLLMRun] = None, **kwargs: Any
+    ) -> Any:
+        """Use tenacity to retry the completion call."""
+        retry_decorator = _create_retry_decorator(self, run_manager=run_manager)
+
+        @retry_decorator
+        def _completion_with_retry(**kwargs: Any) -> Any:
+            try:
+                request_timeout = kwargs.pop("request_timeout", DEFAULT_TIME_OUT)
+                data = kwargs.pop("data")
+                stream = kwargs.pop("stream", self.streaming)
+
+                request = Requests(
+                    headers=self._headers(), auth=self.auth.get("signer")
+                )
+                response = request.post(
+                    url=self.endpoint,
+                    data=data,
+                    timeout=request_timeout,
+                    stream=stream,
+                    **kwargs,
+                )
+                self._check_response(response)
+                return response
+            except TokenExpiredError as e:
+                raise e
+            except Exception as err:
```

Comment:
This will not show us real error form the endpoint? For example if i pass `top_p=-1`, will the error show me that top_p is incorrect?

Reply:
I believe it will show. When parameter is incorrect, the server won't return 200 status code and thus the method `self._check_response(response)` will raise the error, which will show real error form the endpoint. 

e.g.
```
RuntimeError: Error occurs by inference endpoint: Server error: 404 Client Error: Not Found for url: https://modeldeployment.us-ashburn-1.oci.customer-oci.com/<>/predict. 
Message: {"object":"error","message":"The model `llm` does not exist.","type":"invalid_request_error","param":null,"code":null}
```